### PR TITLE
[CSM] Add `on_joinplayer` and `on_leaveplayer` callbacks.

### PR DESCRIPTION
**Pull request by red-001**
Edit: not sure if this is the best way to implement this.

### CODE REVIEW

File: src/content_cao.cpp [hunk 1]
```diff
@@ -667,6 +672,8 @@ void GenericCAO::processInitData(const std::string &data)
 GenericCAO::~GenericCAO()
 {
 	if (m_is_player) {
+		if (m_client->moddingEnabled() && !m_client->isShutdown())
+			m_client->getScript()->on_leaveplayer(m_name);
```
Comment (nerzhul) [inline src/content_cao.cpp]: this is a little bit incorrect as everything is a CAO client side, you distringuish players from other entities

**Comment by rubenwardy** [comment]
Where are GenericCAO initialised and destroyed? Maybe that would b a better candidate for where to place the code

**Comment by nerzhul** [comment]
it's in environment, but here we need to know a player is added and in environment we are not very aware about that (atm)

File: doc/client_lua_api.md [hunk 2]
```diff
@@ -672,6 +672,13 @@ Call these functions only at load time!
     * Called when the local player punches a node
     * Newest functions are called first
     * If any function returns true, the punch is ignored
+* `minetest.register_on_joinplayer(func(name))`
+    * Called when a player joins the game.
+    * Gets called for all players currently on a server when joining a server.
+* `minetest.register_on_leaveplayer(func(name))`
+    * Called when a player leaves the game.
+    * Not called if the client is exiting.
```
Comment (SmallJoker) [inline doc/client_lua_api.md]: May needs some tweaking: `Not called if the own client is exiting.` ?
Otherwise :+1:

Comment (sofar) [inline doc/client_lua_api.md]: Just write
```
 * Called when a remote player connects to | disconnects from the server
```

**Comment by red-001** [comment]
Closing this as implementing it this way means that the callbacks aren't executed if a player joins but is outside of the player transfer distance. This also means that when the client connects the callback is called for all players within that distance that are already connect

**Pull request closed without merging**